### PR TITLE
PureScript method generation

### PR DESCRIPTION
**Pull request by rtrvrtg**
PureScript methods are now generated to abstract away the AJAX call. This also incorporates a bunch of variable-handling tweaks including forcing JSON transport for all requests.

Two examples are provided: a port of the simple counter example from servant-jquery, and a persistent to do list.

### CODE REVIEW

File: examples/counter/temp/api.purs [hunk 1]
```diff
@@ -0,0 +1,70 @@
+module App where
```
Comment (christian-marie) [inline examples/counter/temp/api.purs]: Why is there a temp file here?

File: examples/counter/www/index.html [hunk 2]
```diff
@@ -15,25 +15,40 @@
...
+        $('#counter').html('Counter: ' + response.value);
+      } 
+    }  
+  }
+};
+
 $(document).ready(function() {
   // we get the current value stored by the server when the page is loaded
-  getcounter(updateCounter, alert);
+  PS.App.getcounter(updateCounter)(onError)();
```
Comment (christian-marie) [inline examples/counter/www/index.html]: snake casing is going to be pretty important with the method naming, i.e. getCounter

File: examples/temp/api.purs [hunk 3]
```diff
@@ -1,38 +0,0 @@
-module App where
```
Comment (christian-marie) [inline examples/temp/api.purs]: Temp file again, did you mean that to be here?

File: examples/counter/temp/api.purs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+type Status = String
+type ResponseData = String
+
+type SuccessFn eff = (ResponseData -> Status -> XHR -> (Eff (xhr :: XHREff | eff) Unit))
+type FailureFn eff = (XHR -> Status -> ResponseData -> (Eff (xhr :: XHREff | eff) Unit))
+
+
+type PostcounterHeaders =  { content_Type :: String, accept :: String }
+
+postcounter :: forall eff. (SuccessFn eff) -> (FailureFn eff) -> (Eff (xhr :: XHREff | eff) Unit)
```
Comment (christian-marie) [inline examples/counter/temp/api.purs]: If this is autogenerated, perhaps we don't need all the weird where bindings?

File: examples/todo/src/App/App.Ajax.purs [hunk 5]
```diff
@@ -0,0 +1,103 @@
+module App.Ajax where
```
Comment (christian-marie) [inline examples/todo/src/App/App.Ajax.purs]: Did you meant to commit *more* autogenerated code?

File: examples/todo/src/Data/List/Utils.purs [hunk 6]
```diff
@@ -0,0 +1,70 @@
+module Data.List.Utils (
```
Comment (christian-marie) [inline examples/todo/src/Data/List/Utils.purs]: Surely these are in a library somewhere?

File: examples/todo/src/Data/List/Utils.purs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+) where
+
+import qualified Data.Array as A
+import Data.List
+import Data.Tuple
+
+-- * Some list comparisons
+-- All nicked from MissingH
+
+join :: forall a. [a] -> [[a]] -> [a]
```
Comment (christian-marie) [inline examples/todo/src/Data/List/Utils.purs]: This for instance is intercalate, it appears.

File: examples/counter/temp/api.purs (hunk 1 above)
Comment (rtrvrtg) [inline examples/counter/temp/api.purs]: The temp file is where the intermediary Purescript file is rendered to. We could remove that altogether, but it's probably more useful to see what servant-purescript compiles to on the way.

For the sake of clarity, I might rename that directory to `src` or something and drop other Purescript application code in there.

File: examples/counter/www/index.html (hunk 2 above)
Comment (rtrvrtg) [inline examples/counter/www/index.html]: This is a limitation of servant-jquery itself – the method name comes through entirely in lowercase. Might see about putting in a PR for this.

File: examples/temp/api.purs (hunk 3 above)
Comment (rtrvrtg) [inline examples/temp/api.purs]: Pretty sure I removed all top-level example code. Will check again.

File: examples/todo/src/Data/List/Utils.purs (hunk 6 above)
Comment (rtrvrtg) [inline examples/todo/src/Data/List/Utils.purs]: There doesn't seem to be a library of common list manipulators in a library that's easy to get at. They seem to be scattered all over the place. Some of them are simple synonyms, like `join` & `intercalate` in Data.Foldable, but definitely not anything as extensive and as centralised as Data.List.Utils from Haskell's MissingH.

**Pull request merged**